From 761ceecd3d6d8e864935d96bb9b6f159e8695817 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Thu, 13 Oct 2016 18:26:30 +0800
Subject: [PATCH 1184/5242] MLK-13333-3 ARM: imx: enable bus clock auto gating
 for i.mx6sll

commit  1df5d708aafba40bd05a36370c8797194d98a779 from
https://source.codeaurora.org/external/imx/linux-imx.git

i.MX6SLL has new hardware function of bus auto clock gating,
whenerve bus is idle, its clock will be auto gated to save
power, enable this function.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/pm-imx6.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/arm/mach-imx/pm-imx6.c b/arch/arm/mach-imx/pm-imx6.c
index 56341fb..83a8f82 100644
--- a/arch/arm/mach-imx/pm-imx6.c
+++ b/arch/arm/mach-imx/pm-imx6.c
@@ -91,6 +91,8 @@
 #define UART_UBRC	0xac
 #define UART_UTS	0xb4
 
+#define IOMUXC_GPR5_CLOCK_AFCG_X_BYPASS_MASK	0xf800
+
 extern unsigned long iram_tlb_base_addr;
 extern unsigned long iram_tlb_phys_addr;
 
@@ -1266,6 +1268,7 @@ void __init imx6dl_pm_init(void)
 void __init imx6sl_pm_init(void)
 {
 	struct device_node *np;
+	struct regmap *gpr;
 
 	if (cpu_is_imx6sll()) {
 		imx6_pm_common_init(&imx6sll_pm_data);
@@ -1273,6 +1276,11 @@ void __init imx6sl_pm_init(void)
 			"/soc/aips-bus@02000000/spba-bus@02000000/serial@02020000");
 		if (np)
 			console_base = of_iomap(np, 0);
+		/* i.MX6SLL has bus auto clock gating function */
+		gpr = syscon_regmap_lookup_by_compatible("fsl,imx6q-iomuxc-gpr");
+		if (!IS_ERR(gpr))
+			regmap_update_bits(gpr, IOMUXC_GPR5,
+				IOMUXC_GPR5_CLOCK_AFCG_X_BYPASS_MASK, 0);
 		return;
 	}
 
-- 
1.7.9.5

